### ISSUE TITLE
Losing Shared Folder

### ISSUE THREAD

**Issue by tknerr** [fgrehm/vagrant-lxc#342]
While updating my [vagrant workflow tests](https://github.com/tknerr/vagrant-workflow-tests) so they can be run via vagrant-lxc I noticed that I'm losing a shared folder with vagrant-lxc which I don't lose with the virtualbox provider.

Steps to reproduce:
```
vagrant plugin install vagrant-lxc
vagrant plugin install vagrant-toplevel-cookbooks
git clone https://github.com/tknerr/sample-infrastructure-repo.git
vagrant up app_v1
vagrant provision app_v1
```

With virtualbox provider everything works, with the lxc provider the last `vagrant provision` step fails because it lost a shared folder:
```
tkn@dev-box:~/vagrant-workflow-tests/playground/sample-infrastructure-repo$ vagrant provision app_v1
==> app_v1: Using top-level cookbook 'https://github.com/tknerr/sample-toplevel-cookbook'
==> app_v1: Ensuring top-level cookbook is checked out at 'v0.2.0'
==> app_v1: Installing top-level cookbook dependencies to '/home/tkn/vagrant-workflow-tests/playground/sample-infrastructure-repo/.vagrant/toplevel-cookbooks/app_v1/cookbooks'
Resolving cookbook dependencies...
Fetching 'sample-app' from source at .
Using apache2 (1.10.4)
Using apt (2.4.0)
Using iptables (0.13.2)
Using logrotate (1.6.0)
Using pacman (1.1.1)
Using sample-app (0.2.0) from source at .
Vendoring apache2 (1.10.4) to /home/tkn/vagrant-workflow-tests/playground/sample-infrastructure-repo/.vagrant/toplevel-cookbooks/app_v1/cookbooks/apache2
Vendoring apt (2.4.0) to /home/tkn/vagrant-workflow-tests/playground/sample-infrastructure-repo/.vagrant/toplevel-cookbooks/app_v1/cookbooks/apt
Vendoring iptables (0.13.2) to /home/tkn/vagrant-workflow-tests/playground/sample-infrastructure-repo/.vagrant/toplevel-cookbooks/app_v1/cookbooks/iptables
Vendoring logrotate (1.6.0) to /home/tkn/vagrant-workflow-tests/playground/sample-infrastructure-repo/.vagrant/toplevel-cookbooks/app_v1/cookbooks/logrotate
Vendoring pacman (1.1.1) to /home/tkn/vagrant-workflow-tests/playground/sample-infrastructure-repo/.vagrant/toplevel-cookbooks/app_v1/cookbooks/pacman
Vendoring sample-app (0.2.0) to /home/tkn/vagrant-workflow-tests/playground/sample-infrastructure-repo/.vagrant/toplevel-cookbooks/app_v1/cookbooks/sample-app
==> app_v1: Configuring Chef Solo provisioner for top-level cookbook
==> app_v1: Configuring cache buckets...
==> app_v1: Chef 12.0.3 Omnibus package is already installed.
==> app_v1: Running provisioner: chef_solo...
==> app_v1: Detected Chef (latest) is already installed
Generating chef JSON and uploading...
==> app_v1: Running chef-solo...
==> app_v1: stdin: is not a tty
==> app_v1: [2015-01-18T11:27:50+00:00] INFO: Forking chef instance to converge...
==> app_v1: [2015-01-18T11:27:50+00:00] INFO: *** Chef 12.0.3 ***
==> app_v1: [2015-01-18T11:27:50+00:00] INFO: Chef-client pid: 3213
==> app_v1: [2015-01-18T11:27:51+00:00] INFO: Setting the run_list to ["recipe[sample-app]"] from CLI options
==> app_v1: [2015-01-18T11:27:51+00:00] INFO: Run List is [recipe[sample-app]]
==> app_v1: [2015-01-18T11:27:51+00:00] INFO: Run List expands to [sample-app]
==> app_v1: [2015-01-18T11:27:51+00:00] INFO: Starting Chef Run for appv1.local
==> app_v1: [2015-01-18T11:27:51+00:00] INFO: Running start handlers
==> app_v1: [2015-01-18T11:27:51+00:00] INFO: Start handlers complete.
==> app_v1: [2015-01-18T11:27:51+00:00] ERROR: Running exception handlers
==> app_v1: [2015-01-18T11:27:51+00:00] ERROR: Exception handlers complete
==> app_v1: [2015-01-18T11:27:51+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
==> app_v1: [2015-01-18T11:27:51+00:00] ERROR: Cookbook sample-app not found. If you're loading sample-app from another cookbook, make sure you configure the dependency in your metadata
==> app_v1: [2015-01-18T11:27:51+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
Chef never successfully completed! Any errors should be visible in the
output above. Please fix your recipes so that they properly complete.
```

The reason is that the shared folder `/tmp/vagrant-toplevel-cookbooks/63b3b62fe28610b24b8b8c1ee5cb1f46/cookbooks` somehow gets lost between the initial `vagrant up` and the subsequent `vagrant provision`:
```
...
==> app_v1: Configuring Chef Solo provisioner for top-level cookbook
==> app_v1: Importing base box 'fgrehm/precise64-lxc'...
==> app_v1: Setting up mount entries for shared folders...
    app_v1: /vagrant => /home/tkn/vagrant-workflow-tests/playground/sample-infrastructure-repo
    app_v1: /tmp/vagrant-cache => /home/tkn/.vagrant.d/cache/fgrehm/precise64-lxc
    app_v1: /tmp/vagrant-toplevel-cookbooks/63b3b62fe28610b24b8b8c1ee5cb1f46/cookbooks => /home/tkn/vagrant-workflow-tests/playground/sample-infrastructure-repo/.vagrant/toplevel-cookbooks/app_v1/cookbooks
==> app_v1: Starting container...
...
```

After the `vagrant up` it seems to be empty:
```
tkn@dev-box:~/vagrant-workflow-tests/playground/sample-infrastructure-repo$ /opt/vagrant/bin/vagrant ssh -c "ls -la /tmp/vagrant-toplevel-cookbooks/63b3b62fe28610b24b8b8c1ee5cb1f46/cookbooks" app_v1
total 0
Connection to 10.0.3.213 closed.
```

This is how I'm configuring the shared folder from the vagrant-toplevel-cookbooks plugin:
https://github.com/tknerr/vagrant-toplevel-cookbooks/blob/master/lib/vagrant-toplevel-cookbooks/action/clone.rb#L96-97

And this is where the action is hooked in:
https://github.com/tknerr/vagrant-toplevel-cookbooks/blob/master/lib/vagrant-toplevel-cookbooks/plugin.rb#L28-30 

Any ideas what I'm doing wrong?

**Comment by tknerr** [comment]
Gist with full logs here: https://gist.github.com/tknerr/67ffc7c02923e8a9379a

**Comment by tknerr** [comment]
...just run into this again (google led me here -- already forgot that I created that issue initially... ;-))

**Comment by tknerr** [comment]
...and a quick check revealed that it must have something to do with the way the toplevel-cookbooks mounts the shared folder, because the other shared folders (`/vagrant` and `/tmp/vagrant-cache`) are still functional.

Might be a permission problem:
```
vagrant@appv1:~$ ls -la /tmp/vagrant-toplevel-cookbooks/63b3b62fe28610b24b8b8c1ee5cb1f46/cookbooks/
total 32
drwxr-xr-x 8 1001 users 4096 Feb  4 13:21 .
drwxr-xr-x 3 root root  4096 Feb  4 13:21 ..
drwxr-xr-x 7 1001 users 4096 Feb  4 13:21 apache2
drwxr-xr-x 9 1001 users 4096 Feb  4 13:21 apt
drwxr-xr-x 5 1001 users 4096 Feb  4 13:21 iptables
drwxr-xr-x 7 1001 users 4096 Feb  4 13:21 logrotate
drwxr-xr-x 6 1001 users 4096 Feb  4 13:21 pacman
drwxr-xr-x 7 1001 users 4096 Feb  4 13:21 sample-app
```

vs.

```
vagrant@appv1:~$ ll /tmp/vagrant-cache/
total 40
drwxrwxrwx 8    1001 users    4096 Feb  4 11:28 ./
drwxrwxrwt 4 root    root     4096 Feb  4 14:17 ../
drwxrwxr-x 3 vagrant vagrant  4096 Feb  4 11:22 apt/
drwxrwxr-x 3 vagrant vagrant 12288 Feb  4 13:22 apt_lists/
drwxrwxr-x 9 vagrant vagrant  4096 Feb  4 11:22 busser/
drwxrwxr-x 2 vagrant vagrant  4096 Feb  4 11:29 chef/
drwxrwxr-x 3 vagrant vagrant  4096 Feb  4 11:28 chef_gem/
drwx------ 2 root    root     4096 Feb  4 11:21 vagrant_omnibus/
```

**Comment by tknerr** [comment]
Update: just compared with what [vagrant-berkshelf](https://github.com/berkshelf/vagrant-berkshelf) plugin does:

 * in fact they end up with the same permissions in the shared folders
 * but they *survive* the following `vagrant provision` and the shared folder is *not* lost

=> so unlikely that it's a permission problem

**Comment by tknerr** [comment]
gotcha. this is the culprit:
https://github.com/tknerr/vagrant-toplevel-cookbooks/blob/51727ddc89581bc0be5dd8e1a2cf651164e31473/lib/vagrant-toplevel-cookbooks/action/clone.rb#L84

**Comment by tknerr** [comment]
Fixed with vagrant-toplevel-cookbooks 0.2.4

(was not a bug in vagrant-lxc)
